clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -447,12 +582,40 @@ private boolean handleKafkaRequest(byte[] requestBytes) throws IOException, Auth
                 throw e;
         }
         if (clientMechanism != null) {
+            if (reauthInfo != null)
+                ensureSaslMechanismUnchanged(clientMechanism, requestContext);
             createSaslServer(clientMechanism);
             setSaslState(SaslState.AUTHENTICATE);
         }
         return isKafkaRequest;
     }
 
+    private void ensurePrincipalUnchanged() throws SaslAuthenticationException {
+        KafkaPrincipal reauthenticatedKafkaPrincipal = principal();
+        if (!reauthInfo.previousKafkaPrincipal.equals(reauthenticatedKafkaPrincipal)) {
+            throw new SaslAuthenticationException(String.format(
+                    "Cannot change principals during re-authentication from %s.%s: %s.%s",
+                    reauthInfo.previousKafkaPrincipal.getPrincipalType(), reauthInfo.previousKafkaPrincipal.getName(),
+                    reauthenticatedKafkaPrincipal.getPrincipalType(), reauthenticatedKafkaPrincipal.getName()));
+        }
+    }
+
+    private void ensureSaslMechanismUnchanged(String clientMechanism, RequestContext requestContext)
+            throws UnsupportedSaslMechanismException {
+        if (!reauthInfo.previousSaslMechanism.equals(clientMechanism)) {
+            LOG.debug(
+                    "SASL mechanism '{}' requested by client is not supported for re-authentication of mechanism '{}'",
+                    clientMechanism, reauthInfo.previousSaslMechanism);
+            buildResponseOnAuthenticateFailure(requestContext, new SaslHandshakeResponse(
+                    Errors.UNSUPPORTED_SASL_MECHANISM, new HashSet<>(Arrays.asList(reauthInfo.previousSaslMechanism))));
+            throw new UnsupportedSaslMechanismException("Unsupported SASL mechanism " + clientMechanism);
+        }
+    }
+
+    private String authenticationOrReauthenticationText() {
```

Comment:
This can be in `reauthInfo` as well. And `ReauthInfo` can have a method `authenticating()` or `reauthenticating()` to avoid all the checks for `null`. Same for ClientAuthenticator as well.

Reply:
👍 I moved `authenticationOrReauthenticationText()` into `AuthInfoForReauth` and added a `private boolean reauthenticating()` method to `SaslServerAuthenticator`.

Reply:
Did the same thing for `SaslClientAuthenticator`.